packages/api-fetch/src/middlewares/fetch-all-middleware.js

```diff
@@ -0,0 +1,98 @@
+/**
+ * WordPress dependencies
+ */
+import { addQueryArgs } from '@wordpress/url';
+
+/**
+ * Internal dependencies
+ */
+import apiFetch from '../';
+
+// Apply query arguments to both URL and Path, whichever is present.
+const modifyQuery = ( { path, url, ...options }, queryArgs ) => ( {
+	...options,
+	url: url && addQueryArgs( url, queryArgs ),
+	path: path && addQueryArgs( path, queryArgs ),
+} );
+
+// Duplicates parsing functionality from apiFetch.
+const parseResponse = ( response ) => response.json ?
+	response.json() :
+	Promise.reject( response );
+
+const parseLinkHeader = ( linkHeader ) => {
+	if ( ! linkHeader ) {
+		return {};
+	}
+	const match = linkHeader.match( /<([^>]+)>; rel="next"/ );
+	return match ? {
+		next: match[ 1 ],
+	} : {};
+};
+
+const getNextPageUrl = ( response ) => {
+	const { next } = parseLinkHeader( response.headers.get( 'link' ) );
+	return next;
+};
+
+const requestContainsUnboundedQuery = ( options ) => {
+	const pathIsUnbounded = options.path && options.path.indexOf( 'per_page=-1' > -1 );
+	const urlIsUnbounded = options.url && options.url.indexOf( 'per_page=-1' > -1 );
+	return pathIsUnbounded || urlIsUnbounded;
+};
+
+// The REST API enforces an upper limit on the per_page option. To handle large
+// collections, apiFetch consumers can pass `per_page=-1`; this middleware will
+// then recursively assemble a full response array from all available pages.
+const fetchAllMiddleware = async ( options, next ) => {
+	if ( options.parse === false ) {
+		// If a consumer has opted out of parsing, do not apply middleware.
+		return next( options );
```

Comment:
Should we throw an error here if we detect `per_page: -1` in the request as the server won't know how to deal with it?

Reply:
I feel like we should let it through; we don't do that anywhere ourselves, and the server itself will throw the error, which should make the issue pretty clear.